onnx/shape_inference/implementation.cc

```diff
@@ -223,245 +193,283 @@ void mergeShapesAndTypes(const TypeProto& inferredType, TypeProto* existingType)
     }
     // Support IR>=4: some tensors can only exist in initializer and not in input
     // So shape_inference should make use of initializer shapes
     // Store initializer shape info in value_info as well
     else if (ir_version >= 4) {
-      initializerTypeList.push_back(std::move(initializerType));
-      valueTypesByName[name] = &initializerTypeList.back();
+      initializer_type_list.push_back(std::move(initializer_type));
+      value_types_by_name[name] = &initializer_type_list.back();
     }
   }
 
   bool has_experimental_op = false;
-  // If encounter experimental op, stop checking
+  // Collect data from constant nodes and check if any experimental ops exist
   for (const auto& n : g->node()) {
     if (checker::check_is_experimental_op(n.op_type())) {
-      std::cerr << "Warning: Shape inference does not support"
-                << " models with experimental operators: " << n.op_type() << std::endl;
       has_experimental_op = true;
-    }
-  }
-
-  // Collect data from constant nodes.
-  for (const auto& n : g->node()) {
-    if (n.op_type() != "Constant" || n.output().size() != 1) {
-      continue;
-    }
-    for (const auto& attr : n.attribute()) {
-      if (attr.name() == "value") {
-        if (attr.type() == AttributeProto::TENSOR && attr.has_t()) {
-          inputDataByName[n.output(0)] = &attr.t();
-        } else if (attr.type() == AttributeProto::SPARSE_TENSOR && attr.has_sparse_tensor()) {
-          inputSparseDataByName[n.output(0)] = &attr.sparse_tensor();
+    } else if (n.op_type() == "Constant" &&  n.output().size() == 1) {
+      for (const auto& attr : n.attribute()) {
+        if (attr.name() == "value") {
+          if (attr.type() == AttributeProto::TENSOR && attr.has_t()) {
+            input_data_by_name[n.output(0)] = &attr.t();
+          } else if (attr.type() == AttributeProto::SPARSE_TENSOR && attr.has_sparse_tensor()) {
+            input_sparse_data_by_name[n.output(0)] = &attr.sparse_tensor();
+          }
         }
       }
     }
   }
 
   std::vector<std::string> inference_errors;
   bool has_unsupported_op = false; // check whether exist unsupported ops
-
   for (auto& n : *g->mutable_node()) {
     // Resolve domain for node
     auto dit = opset_imports.find(n.domain());
     if (dit == opset_imports.end()) {
-      continue;
+      fail_type_inference("Cannot infer type and shape for node name ", n.name(), ". No opset import for domain",
+          n.domain(), " optype ", n.op_type());
     }
     auto domain_version = dit->second;
     const auto schema = schema_registry->GetSchema(n.op_type(), domain_version, n.domain());
-    InferenceContextImpl ctx(n, valueTypesByName, inputDataByName,inputSparseDataByName,
-      &generatedShapeDataByName, &graphInferenceContext);
-    if (!schema) {
-      std::cerr << "Warning: Unsupported operator " << n.op_type() << ". No schema registered for this operator."
-                << std::endl;
```

Comment:
So ONNX finally decides not to print such warnings anymore (due to https://github.com/onnx/onnx/issues/3696 and https://github.com/onnx/onnx/issues/2239)?

I am afraid without those warnings, users will complain the shape inference is incomplete and there is no any warning/error... Could ONNX at least throw a warning about unsupported error when enabling strict_mode?

Reply:
Or perhaps ONNX can add a new "mode" for shape_inference to detect such warning/error (unsupported schema)